Restructure AIP-193

jskeet opened the pull request.

jskeet committed:
AIP-193: Move JSON representation details to the bottom, and clarify

This is more of an informative section than anything else - API designers only need to know about it in order to understand the JSON response they might see from curl etc.

jskeet committed:
AIP-193: First part of restructuring the description of Status fields

(This doesn't include anything about the details "payloads" yet.)

jskeet committed:
Rewrite ErrorInfo description

Note: this has removed the term "dynamic variables" in favor of only referring to it as "metadata". It's handy to have a single well-defined term, and given that the field name is `metatadata`, I've kept that.

noahdietz commented on aip/general/0193.md:
```diff
@@ -34,272 +34,154 @@ is necessary, you **should** provide a link where a reader can get more
 
-A JSON representation of an error response might look like the following
-example.
+The `message` field is a a developer-facing, human-readable "debug message"
```
```suggestion
The `message` field is a developer-facing, human-readable "debug message"
```
extra `a`

jskeet replied:
Done.

noahdietz commented on aip/general/0193.md:
```diff
@@ -34,272 +34,154 @@ is necessary, you **should** provide a link where a reader can get more
-example.
+The `message` field is a a developer-facing, human-readable "debug message"
+which **should** be in English. (Localized messages are expressed using
+a `LocalizedMessage` within the `details` field.) Any dynamic aspects of
```
```suggestion
a `LocalizedMessage` within the `details` field - see [`Status.details`](#status-details).) Any dynamic aspects of
```
Probably worth linking to? I think this link will work...

jskeet replied:
I think it's better to link to the LocalizedMessage part of the doc (which I haven't updated yet). Adding the link speculatively.

noahdietz replied:
Ah fair point. Thanks

noahdietz commented on aip/general/0193.md:
```diff
@@ -34,272 +34,154 @@ is necessary, you **should** provide a link where a reader can get more
+The `message` field is a a developer-facing, human-readable "debug message"
+which **should** be in English. (Localized messages are expressed using
+a `LocalizedMessage` within the `details` field.) Any dynamic aspects of
+the message **must** be included as metadata within the `ErrorInfo` message
```
```suggestion
the `message` **must** be included as metadata within the `ErrorInfo` message
```
Nit: consistently format as code when referring to the `field`?

jskeet replied:
Let's talk about this a bit more - I tend to find that it's worth keeping things not in code font when they're not *explicitly* talking about the field (which this isn't - it's talking about the semantic meaning). I'm happy to be persuaded, and I suspect there will be some areas where I'm inconsistent already... but to my eyes, a page where a huge proportion of the text is in code font ends up being hard to read.

noahdietz replied:
I agree with your position 100%. I think I was getting confused with semantic meaning vs. concrete reference reading the document top to bottom.

noahdietz replied:
That said, my personal tendency is to explicitly state "`message` value" or "the value of `message`" or "the content of `message`" rather than depend on context clues (like code formatting) to do that "comprehension lifting". I'm not too fussed about it though.

jschneider11 replied:
Agreed. I'd only put the word "message" in quotes when you're referring directly to the field.

noahdietz commented on aip/general/0193.md:
```diff
@@ -34,272 +34,154 @@ is necessary, you **should** provide a link where a reader can get more
+The `message` field is a a developer-facing, human-readable "debug message"
+which **should** be in English. (Localized messages are expressed using
+a `LocalizedMessage` within the `details` field.) Any dynamic aspects of
+the message **must** be included as metadata within the `ErrorInfo` message
```
```suggestion
the message **must** be included as metadata within the `ErrorInfo` that appears
```
Sorry, I get tripped up reading "`ErrorInfo` message" when there is no `message` in `ErrorInfo` - I know you are referring to the proto `message`, but in line I think it is a little confusing. Maybe just refer to the `ErrorInfo` as a singular thing that will be in the `details` as I've suggested?

jskeet replied:
Yup, that definitely works (and I completely understand the tripping up).

noahdietz replied:
Thanks!

noahdietz commented on aip/general/0193.md:
```diff
@@ -34,272 +34,154 @@ is necessary, you **should** provide a link where a reader can get more
+which **should** be in English. (Localized messages are expressed using
+a `LocalizedMessage` within the `details` field.) Any dynamic aspects of
+the message **must** be included as metadata within the `ErrorInfo` message
+in `details`.
```
```suggestion
in [`Status.details`](#status-details).
```
Linking for linking sake - again, I hope it works.

jskeet replied:
Done (and because I'm linking to LocalizedMessage before, it doesn't break my personal semi-rule of avoiding linking to the same thing twice in quick succession!)

noahdietz replied:
Good rule. Thanks!

noahdietz commented on aip/general/0193.md:
```diff
@@ -34,272 +34,154 @@ is necessary, you **should** provide a link where a reader can get more
-error response into sections. Note that the order in which you write the
-error message is often different from how the error is presented to
-users.
+The message is considered a problem description. It is intended for
```
```suggestion
The `message` is considered a problem description. It is intended for
```
Nit: format as code when referring to the field

jskeet replied:
Not done yet, for reason above. (If it had been "The value of message" I would have agreed - but this is the concept rather than the field.) But let's leave this conversation so it's easy to find and change if we go with "make it look like code anywhere that it *could* refer to the field.

noahdietz commented on aip/general/0193.md:
```diff
@@ -34,272 +34,154 @@ is necessary, you **should** provide a link where a reader can get more
-users.
+The message is considered a problem description. It is intended for
+developers to understand the problem and is more detailed than
+[`ErrorInfo.reason`][ErrorInfo-reason], discussed in the next section.
```
```suggestion
[`ErrorInfo.reason`][ErrorInfo-reason], discussed later.
```
The next section is actually about the `Status.code`, so maybe just say `later` and perhaps link to the subheading?

The `[ErrorInfo-reason]` link goes to the googleapis proto btw.

jskeet replied:
Works for me.

noahdietz commented on aip/general/0193.md:
```diff
@@ -34,272 +34,154 @@ is necessary, you **should** provide a link where a reader can get more
+developers to understand the problem and is more detailed than
+[`ErrorInfo.reason`][ErrorInfo-reason], discussed in the next section.
 
+Messages **should** use simple descriptive language that is easy to understand
```
```suggestion
A `message` **should** use simple descriptive language that is easy to understand
```
Nit: maybe I'm being overzealous here but "message" is so overloaded I get pedantic about how we refer to it.

jskeet replied:
Another "let's discuss how much we put in code". (I'm absolutely fine with being pedantic, btw - I'm just being pedantic in a different direction. Once we've agreed a direction, we should nitpick as far as possible.)

noahdietz replied:
I'm ok with what you've done here. Resolving.

noahdietz commented on aip/general/0193.md:
```diff
@@ -34,272 +34,154 @@ is necessary, you **should** provide a link where a reader can get more
+without machine-readable identifiers, the value of `message` **must**
+remain the same for any given error, as developers have previously had no
+option but to use this for error handling. For more information, see
+[Changing Error Messages](#changing-error-messages).
```
Nit: the `as developers have previously...` portion is Rationale and might be better suited in a dedicated Rationale section for reference. That said, I realize you are working with existing text and I'm not too fussed about it.

jskeet replied:
Let's leave this comment thread open for now, as I probably don't have the right place to move this to just yet. But I agree on moving it to the rationale eventually.

jskeet replied:
Ah, now that I've finished the PR, we can potentially action this. I suspect I can just remove:

> as developers have previously had no option but to use this for error handling.

(And keep the link in the following sentence.) Does that sound right?

jskeet replied:
Done exactly as stated.

noahdietz commented on aip/general/0193.md:
```diff
@@ -34,272 +34,154 @@ is necessary, you **should** provide a link where a reader can get more
-in the JSON reply would be `400`.
-
-#### [ErrorDetails][details]
+For example, the value 5 is the numeric value of the `NOT_FOUND`
```
```suggestion
For example, the value `5` is the numeric value of the `NOT_FOUND`
```

jskeet replied:
Done.

noahdietz commented on aip/general/0193.md:
```diff
@@ -34,272 +34,154 @@ is necessary, you **should** provide a link where a reader can get more
 
-The following section describes `ErrorInfo`, `LocalizedMessage`, and
-`Help` in more detail.
+The `details` field allows any multiple distinct messages with
```
"any multiple distinct messages" reads a little awkwardly to me. What is this trying to say? "any number of distinct messages"?

jskeet replied:
I agree it's awkward. It's basically trying to say "It's a repeated field, but you can't use the same message more than once". I think "message" makes all of this harder as well. (I haven't got a better alternative yet for that part though.)

jskeet replied:
(The distinctness is covered later, so I've just removed it. See what you think.)

noahdietz commented on aip/general/0193.md:
```diff
@@ -34,272 +34,154 @@ is necessary, you **should** provide a link where a reader can get more
+[Changing Error Messages](#changing-error-messages).
+
+Most details types are optional, but all error responses **must** include
+an `ErrorInfo` message within `details`.
```
```suggestion
All error responses **must** include an `ErrorInfo` in the `details`.
```
I'm just not sure what value the "most are optional" language provides. Could just be me though.

jskeet replied:
That's fine - removed.

noahdietz commented on aip/general/0193.md:
```diff
@@ -34,272 +34,154 @@ is necessary, you **should** provide a link where a reader can get more
+
+#### ErrorInfo
 
 The [`ErrorInfo`][ErrorInfo] message is the required way to send a
```
```suggestion
The [`ErrorInfo`][ErrorInfo] message is the primary way to send a
```

I know it's existing text, but....kind of want to keep the requirements in AIP-style requirements words, ya know?

jskeet replied:
Fine by me. Done.

noahdietz commented on aip/general/0193.md:
```diff
@@ -34,272 +34,154 @@ is necessary, you **should** provide a link where a reader can get more
 
-**Note:** `ErrorInfo` represents a special case. There **must** be
-exactly one `ErrorInfo`. It is required.
+The reason should be terse, but meaningful enough for a human reader to
```
```suggestion
The `reason` **should** be terse, but meaningful enough for a human reader to
```

jskeet replied:
(Same discussion as elsewhere about fields vs concepts.)

noahdietz commented on aip/general/0193.md:
```diff
@@ -34,272 +34,154 @@ is necessary, you **should** provide a link where a reader can get more
-**Note:** `ErrorInfo` represents a special case. There **must** be
-exactly one `ErrorInfo`. It is required.
+The reason should be terse, but meaningful enough for a human reader to
+understand what the reason refers to.
```
```suggestion
understand what the `reason` refers to.
```

jskeet replied:
(And again.)

noahdietz replied:
I agree with your approach and Julie +1'd it. Resolving.

noahdietz commented on aip/general/0193.md:
```diff
@@ -34,272 +34,154 @@ is necessary, you **should** provide a link where a reader can get more
-Without such context, it is unlikely that the message will be fully
-actionable by the user.
+The `domain` field is the logical grouping to which the `reason` belongs.
+The error domain  typically the registered service name of the tool or product
```
```suggestion
The error domain is typically the registered service name of the tool or product
```

jskeet replied:
I hadn't even seen this sentence - I've somewhat repeated it later. Changed this whole paragraph.

jskeet replied:
(I don't think "tool or product" is particularly useful here. Clients are talking to an API - what kind of "tool" is relevant?)

noahdietz commented on aip/general/0193.md:
```diff
@@ -34,272 +34,154 @@ is necessary, you **should** provide a link where a reader can get more
-actionable by the user.
+The `domain` field is the logical grouping to which the `reason` belongs.
+The error domain  typically the registered service name of the tool or product
+that  generated the error. The domain must be a globally unique value, and is
```
```suggestion
that  generated the error. The domain **must** be a globally unique value, and is
```

jskeet replied:
Done.

noahdietz commented on aip/general/0193.md:
```diff
@@ -34,272 +34,154 @@ is necessary, you **should** provide a link where a reader can get more
-This practice is critical so that machine actors do not need to rely on
-`LocalizedMessage.message`, which is subject to change and is not part
-of the API contract.
+The (reason, domain) pair form a machine-readable way of identifying a particular error.
```
```suggestion
The (`reason`, `domain`) pair form a machine-readable way of identifying a particular error.
```
Here and elsewhere

jskeet replied:
Same response :)

noahdietz replied:
I agree with your approach and Julie +1'd it. Resolving.

noahdietz commented on aip/general/0193.md:
```diff
@@ -34,272 +34,154 @@ is necessary, you **should** provide a link where a reader can get more
+Services **must** use the same (reason, domain) pair for the same error, and
+**must not** use the same (reason, domain) pair for logically different errors.
+The decision about whether two errors are "the same" or not is not always clear, but
+should generally be considered in terms of the expected action a client might take
```
```suggestion
**should** be considered in terms of the expected action a client might take
```

jskeet replied:
Done.

noahdietz commented on aip/general/0193.md:
```diff
@@ -34,272 +34,154 @@ is necessary, you **should** provide a link where a reader can get more
-> East".  It is expected to be available again on 2199-05-13.
+Any request-specific information which contributes to the `Status.message` or
+`LocalizedMessage.message` messages **must** be represented within `metadata`.
+This practice is critical so that machine actors do not need to parse error
+messages to extract information.
```
Nit: similar to earlier, this reasoning would be best in a Rationale section but I'm not going to be picky

jskeet replied:
Let's leave this open so we can move it and link.

jschneider11 replied:
Agreed.

jskeet replied:
I haven't found a good way of doing this. I'm tempted to suggest we get this PR in but raise an issue to review the new text carefully for rationale-like text - I suspect there's a fair amount of it here. Personally I'm not *hugely* against it in this particular case, mind you...

noahdietz replied:
That's fine with me (punting)

noahdietz commented on aip/general/0193.md:
```diff
@@ -34,272 +34,154 @@ is necessary, you **should** provide a link where a reader can get more
-*required* `ErrorInfo`). For example, the `metadata` map for the sample
-error message above will include *at least* the following key/value
-pairs:
+> An &lt;e2-medium&gt; VM instance with &lt;local-ssd=3,nvidia-t4=2&gt; is currently unavailable
+> in the &lt;us-east1-a&gt; zone. Consider trying your request in the &lt;us-central1-f,us-central1-c&gt;
+> zone(s), which currently has/have capacity to accommodate your request. Alternatively,
+> you can try your request again with a different VM hardware configuration
+> or at a later time. For more information, see the troubleshooting documentation.
```
I think there was some copy paste format issues here with the `&lt;e2-medium&gt;` stuff

jskeet replied:
Nope, this was deliberate. That's how I'd expect to represent `<` and `>` in Markdown when they're not in code. I don't know what the AIP site generator will do, but it's deliberately as it is :)

noahdietz replied:
TIL! let's see what happens :)

jschneider11 replied:
I suggest checking how these HTML characters render. I'm showing a blank for the rendered zone name in the following text:

Consider trying your request in the , zone(s), which currently has/have capacity to accommodate your request. Alternatively, you can try your request again with a different VM hardware configuration or at a later time.

jskeet replied:
Basically I don't know what the site generator will do - it's possible that backslashes are the way forward.

@noahdietz: How easy is it run the site generator, and will it involve installing a load of things? (If you have it set up locally and could test this, that would be really handy.)

noahdietz replied:
@jskeet there are instructions on how to run it locally here: https://google.aip.dev/contributing#local-installation

Though I haven't been able to get it running myself - not sure if it is a python version thing or not. It's a black box to me tbh

jskeet replied:
I've actually got it working now - and it looks like using the HTML escaping works, whereas backslashes don't. Once we've merged, we should immediately double check though...

noahdietz replied:
Awesome! SGTM, resolving

noahdietz commented on aip/general/0193.md:
```diff
@@ -34,272 +34,154 @@ is necessary, you **should** provide a link where a reader can get more
+- "zone": "us-east1-a",
+- "vmType": "e2-medium",
+- "attachment": "local-ssd=3,nvidia-t4=2",
+- "zonesWithCapacity": "us-central1-f,us-central1-c"
```
Should these be formatted as code? I guess wondering why it was switched to bulleted list even though it is a "code examle"?

jskeet replied:
I'm not sure... it's not *really* code - they're name/value pairs. (I replaced the example entirely, as I thought it was unrealistic - if we're going to use books, I wouldn't expect a title, but a resource name... I figured using the same example that we have for the complete error would be better.)

The commas were a mistake though.

I've reformatted each entry in the map as code - we could reformat it all as a JSON object if you'd like?

noahdietz commented on aip/general/0193.md:
```diff
@@ -34,272 +34,154 @@ is necessary, you **should** provide a link where a reader can get more
-map for the error payload to be backwards compatible, even if the value
-for a particular key is empty. Keys **must** be expressed as lower
-camel-case.
+In other words, once a user has observed a given key for a (reason, domain) pair they should be
```
```suggestion
In other words, once a user has observed a given key for a (reason, domain) pair they **must** be
```
stronger guidance? If not, perhaps just bold the `should`...or change to something else if you want to just make an assertion e.g. `the need to be able to rely...`

jskeet replied:
Rewritten - I tend to think that every "must" and "should" ought to be aimed at the target audience of the AIP - usually the service owner, rather than the user. (Client library AIPs are slightly different.) Each should/must ought to be effectively an instruction - whereas saying "the user must be able to do something" isn't really an instruction to the subject of the verb... see what you think of this rewrite :)

jschneider11 replied:
LGTM.

noahdietz commented on aip/general/0193.md:
````diff
@@ -458,6 +340,85 @@ If the user does have proper permission, but the requested resource or
+```json
+{
+  "error": {
+    "code": 404,          // The HTTP status code for "not found"
````
How did you get coloring in here? Cool - not sure if it will render in AIP site generator

jskeet replied:
That's just GitHub saying "JSON doesn't really have comments". I think we can probably just see how it renders and fix it later if it looks really bad :)

noahdietz replied:
TIL haha

jskeet reviewed (commented):
I've made the changes within the existing commit structure (I hope).

It's possible that will lose the comments - but I hope not. (If it does, I'll abandon that approach for the rest of the commits.)

jskeet committed:
AIP-193: Update the LocalizedMessage section

Note that we don't currently (as far as I'm aware) have a good standardized way of letting users specify a locale. (Can they use Accept-Language, for example? What about in gRPC?) We can tackle that separately.

This commit changes the guidance in a couple of important ways:

- It says that the locale should be a user-specified one rather than a "service's native language"
- It calls out the ability to use LocalizedMessage even when we don't have a user-specified language, when Status.message can't be changed for compatibility reasons

This commit *doesn't* address the issue of the target audience of LocalizedMessage; later text claims this is aimed at end-users whereas Status.message is more aimed at developers. I'm not convinced by that at the moment.

jskeet committed:
AIP-193: Update the Help section

Significant changes here (all up for discussion):

- Explicitly include Status.message as well as LocalizedMessage.message
- Remove the standardized https://cloud.google.com/PRODUCT/docs/ERROR-REASON format requirement
- Require the description to be suitable as text for a link
- Require the description to be plain text
- Require the URL to be absolute and include a scheme
- Add authentication requirements

noahdietz reviewed (commented):
Your responses to my review and the following two commits lgtm

yordis commented on aip/general/0193.md:
```diff
@@ -458,6 +327,85 @@ If the user does have proper permission, but the requested resource or
+    "message": "The zone 'us-east1-a' does not have enough resources available to fulfill the request. Try a different zone, or try again later.",
+    "status": "RESOURCE_EXHAUSTED",
+    "details": [
+      {
+        "@type": "type.googleapis.com/google.rpc.ErrorInfo",
+        "reason": "RESOURCE_AVAILABILITY",
+        "domain": "compute.googleapis.com",
+        "metadata": {
+          "zone": "us-east1-a",
+          "vmType": "e2-medium",
+          "attachment": "local-ssd=3,nvidia-t4=2",
+          "zonesWithCapacity": "us-central1-f,us-central1-c"
+        }
+      },
```
Is the `ErrorInfo` optional? I wonder why it wouldn't be part of the `error` itself instead of `details`

Likewise with other things like localized message, if they suppose to be only one entry, since `details` is a list of anything so, it could have many error info or localized messages.

Sorry for the confusion.

jskeet replied:
As noted earlier in the AIP:

- There must *always* be an ErrorInfo. (I suspect that changing `Status` itself to include that directly would be hard in various ways.)
- There must only ever be one instance of any given message type. Yes, that's not a constraint that's expressed in the schema itself (because it can't be) but is specified in the text.

yordis replied:
> Yes, that's not a constraint that's expressed in the schema itself (because it can't be) but is specified in the text.

Wouldn't it help the engineers to make such an invariant part of the schema itself? 

There are some breaking changes, so it is not that easy to adopt or change, but I am just curious about the motivation for using `details` for it overall.

jskeet replied:
> Wouldn't it help the engineers to make such an invariant part of the schema itself?

There's no way of saying "This can use arbitrary messages, but only once per message." We could have had a message with all the "known" additional info messages, but that would have been limiting in terms of APIs wanting to add their own information outside these AIPs.

Yes, in an alternative universe we could have made `ErrorInfo` a field in `Status` - but that's not where we are, and that choice is not up for debate.

noahdietz replied:
In a vacuum, @yordis, yes adding a concrete field to the schema would benefit developers, but there are a few points to using `details`...

1. The reality is that `google.rpc.Status` is used in a number of non-Google-specific ways, so shipping further Google-isms in the schema may be undesirable
2. `details` gives us the flexibility to add new properties without versioning `google.rpc.Status` (which would be infeasible given the breadth of usage)
3. `details` gives *anyone* i.e. non-Google users, the flexibility to add new properties without needing to change `google.rpc.Status`

Without a governance model around `google.rpc.Status` similar to that of HTTP protocol, further extending `google.rpc.Status` concretely would be a difficult task. Instead, we invest in our 1p clients offered to customers to present these well-known details in a reasonable way.

yordis replied:
@noahdietz @jskeet, I sincerely appreciate your taking the time to explain a bit more. 

> Yes, in an alternative universe we could have made ErrorInfo a field in Status - but that's not where we are, and that choice is not up for debate.

That is precisely the confirmation I needed. It has been almost half a decade since I started reverse-engineering and tried to understand every significant business's specifications. I took massive inspiration from AIP-193 for https://github.com/straw-hat-team/adr/pull/30 (if you have some spare time, I would sincerely appreciate your participation in it, just a second try in that alternative universe).

I just wasn't confident enough about this particular scenario. Seriously, thank you so much, years trying to figure out the WHY behind it!

noahdietz replied:
For sure, good luck @yordis, cheers!

jskeet reviewed (commented):
Thanks for the reviews so far - I'll ping again when I've added the commits for the final sections. (We may *also* want to separate out "everything to do with Status" from "how to decide which error codes to use, but that can be a later PR.)

jskeet committed:
AIP-193: Update error messages section (including rationale)

Note: this removes the "message alignment" section which I found confusing.
We could restore and edit it should we wish.

jskeet committed:
AIP-193: Update the rationale around LocalizedMessage

Note: this removes the idea that LocalizedMessage is for "end users" as a distinction between that and Status.message being for developers. If we still want that distinction, we should talk about it more and clarify it - in particular, if we're using LocalizedMessage as an alternative for Status.message where the latter can't be changed, they can't really cater for different audiences.

noahdietz commented on aip/general/0193.md:
```diff
@@ -472,41 +392,49 @@ variable information, so there needs to be machine-readable component of
+RPCs which have **always** included `ErrorInfo` are in a better position:
+the contract is then more about the stability of `ErrorInfo` for any given
+error: the reason and domain need to be consistent over time, and the
+metadata provided for any given (reason,domain) can only be expanded.
```
Nit: Is it OK to have multiple colons in the same sentence?

jskeet replied:
No, it's horrible - sorry about that. I think changing the second colon to a period would be better. Will do that locally, but not create a commit just for that. (I'll wait until we get more feedback to avoid a storm of tiny commits.)

noahdietz reviewed (approved):
Overall LGTM

For posterity, we want a few more reviews, specific @shwoodard and @jschneider11

jskeet commented on aip/general/0193.md:
```diff
@@ -34,401 +34,242 @@ is necessary, you **should** provide a link where a reader can get more
 [`PreconditionFailure`][PreconditionFailure].
 
-##### ErrorInfo payload
+Structured details with machine-readable identifiers **must** be used so
```
I'm now considering removing this paragraph entirely. It's already covered later on, and the currnet "if an error message does not have a machine-readable identifier in addition to the message" effectively contradicts the requirement that we always provide an ErrorInfo.

noahdietz replied:
Yeah I'm a little unsure of the value this paragraph is now providing. Agreed that it adds some confusion.

jskeet replied:
I've tacked just "This provides machine-readable identifiers so that that users can write code against specific aspects of the error" onto the next paragraph (which was otherwise very short anyway). I think that captures the important bit.

jskeet commented on aip/general/0193.md:
```diff
@@ -34,401 +34,242 @@ is necessary, you **should** provide a link where a reader can get more
+change. For more information, see
+[Changing Error Messages](#changing-error-messages).
+
+All error responses **must** include an `ErrorInfo` message within `details`.
```
I suspect it's worth removing "message" within this sentence. Done locally (but not pushed yet).

noahdietz replied:
Yes agreed.

jschneider11 reviewed (commented):
I've been requested to review, but can't approve this PR.

LGTM, though.

jskeet committed:
AIP-193 review feedback and add a dated changelog entry

jskeet reviewed (commented):
I've now left a "potentially last commit" - I had to force push, due to (I suspect) some local modifications, but it's only the last commit that should require re-review.

noahdietz commented on aip/general/0193.md:
````diff
@@ -34,401 +34,235 @@ is necessary, you **should** provide a link where a reader can get more
-library: "Garfield East"
-expectedReturnDate: "2199-05-13"
-```
+The reason should be terse, but meaningful enough for a human reader to
````
```suggestion
The reason **should** be terse, but meaningful enough for a human reader to
```

Nit: Is this a requirement word or should we use a different one (or should we not fuss about use of requirement wrods in a non-requirement context?)

jskeet replied:
I'm fine for it to be a requirement word. Added a new comment for this and the other example.

noahdietz commented on aip/general/0193.md:
```diff
@@ -34,401 +34,235 @@ is necessary, you **should** provide a link where a reader can get more
-- However, the `Status.message` *string* **must** not be changed, as
-  this change is backward-incompatible.
+Textual error messages can be present in both `Status.message` and
+`LocalizedMessage.message` fields. Messages should be succinct but
```
```suggestion
`LocalizedMessage.message` fields. Messages **should** be succinct but
```
ditto re:requirement word or not

jskeet committed:
AIP-193: Address review comments
